### PR TITLE
Ttonev/grid excel style editing vnext

### LINKED ISSUE #3219: Add customization section to Grid Cell Editing topic for excel like editing in grid
Add section and a sample to demonstrate how to customize the igx-grid and its editing to have excel like editing

### PR DESCRIPTION
**Pull request by tishko0**
Closes #3219 

### Checklist:
 
 - [ ] check topic's TOC/menu and paragraph headings
 - [ ] link to other topics using `../relative/path.md`
 - [ ] at the References section at the end of the topic add links to topics, samples, etc
 - [ ] reference API documentation instead of adding a section with API

<br />

 - [ ] use valid component names - [Data] Grid, `IgxSelectComponent`, `<igx-combo>`
 - [ ] use spell checker tool (VS Code, Grammarly, Microsoft Editor)
 - [ ] add inline `code blocks` for the names of classes / tags / properties
 - [ ] add language descriptor for the ```code blocks```
 - [ ] check broken links (use browser add-on)
 - [ ] check if sample is working and fully visible in the topic
 - [ ] check if sample is working and fully visible in the StackBlitz
 - [ ] check if code blocks match the code in StackBlitz demo

<br />

 - [ ] follow SEO guidelines to fill topic's metadata ([SEO guidelines](https://infragisticsinc297.sharepoint.com/:w:/g/Groups/marketing/EWz9InT4FDlErHCumxsKGY4Bd8H03yhRWxDFk47luRz-_Q?e=S5wWcx))

<br />

 - [ * ] do not resolve requested changes (leave that to the reviewer)
 - [ ] add `pending-localization` label when the review of the PR is done
 - [ ] add a member from the localization team to translate it

### CODE REVIEW

File: en/components/grids_templates/cell-editing.md [hunk 1]
```diff
@@ -180,6 +180,78 @@ This code is used in the sample below which implements an [`IgxSelectComponent`]
...
+      return  dataView.findLastIndex((rec, index) => index < currentRowIndex && this.isEditableDataRecordAtIndex(index, dataView));
+    }
+    // or for the next one down the field
+    return dataView.findIndex((rec, index) => index > currentRowIndex && this.isEditableDataRecordAtIndex(index, dataView));
+  }
+```
+
+Please check the full sample bellow:
+
+## Angular Grid cell editing and edit templates Example
```
Comment (mddragnev) [inline en/components/grids_templates/cell-editing.md:242]: Please update the title of the sample.

File: en/components/grids_templates/cell-editing.md [hunk 2]
```diff
@@ -180,6 +180,78 @@ This code is used in the sample below which implements an [`IgxSelectComponent`]
...
+           data-demos-base-url="{environment:demosBaseUrl}" 
+           iframe-src="{environment:demosBaseUrl}/grid/grid-editing-excel-style" alt="Angular Grid Excel Style Editing Example">
+</code-view>
+
+
+Main benefits of the above approach include:
+
+- Constant edit mode: typing while a cell is selected will immediately enter it in edit mode with the value typed, replacing the existing one
+- Any non-data rows are skipped when navigating with `Enter`/`Shift+Enter`
+- Should grouping, master detail or any other feature adding headers as records is enabled, enter will navigate you to the next editable cell skipping all headers. Cycling quickly trough the values has neever been easier
```
Comment (mddragnev) [inline en/components/grids_templates/cell-editing.md:254]: Arent the last two with the same meaning. I think that they should be merged somehow. Also, the last part `Cycling quickly trough the values has neever been easier` has a typo and IMO it is not appropriate for the topics.

File: en/components/grids_templates/cell-editing.md [hunk 3]
```diff
@@ -180,6 +180,76 @@ This code is used in the sample below which implements an [`IgxSelectComponent`]
...
+<code-view style="height:550px" 
+           data-demos-base-url="{environment:demosBaseUrl}" 
+           iframe-src="{environment:demosBaseUrl}/grid/grid-editing-excel-style" alt="Angular Grid Excel Style Editing Example">
+</code-view>
+
+
+Main benefits of the above approach include:
+
+- Constant edit mode: typing while a cell is selected will immediately enter it in edit mode with the value typed, replacing the existing one
+- Any non-data rows are skipped when navigating with `Enter`/`Shift+Enter`. Cycling quickly trough the values has neever been easier
```
[the remark this answers is not in the document]
Comment (mddragnev) [reply en/components/grids_templates/cell-editing.md:252]: Also, do you think that it is okay for something like this to be written in the docfx @dkamburov . Talking about the last sentence.

[the remark this answers is not in the document]
Comment (dkamburov) [reply en/components/grids_templates/cell-editing.md:252]: @tishko0 Yes, I agree with @mddragnev here, let's make it a bit more formal(this sounds like commercial).

How about something like this for example:
**Any non-data rows are skipped when navigating with `Enter`/`Shift+Enter`. This allows users to quickly cycle through their values.**

[the remark this answers is not in the document]
Comment (dkamburov) [reply en/components/grids_templates/cell-editing.md:252]: > @tishko0 Yes, I agree with @mddragnev here, let's make it a bit more formal(this is good, but sounds like commercial a bit).
> 
> How about something like this for example: **Any non-data rows are skipped when navigating with `Enter`/`Shift+Enter`. This allows users to quickly cycle through their values.**

**Comment by dkamburov** [comment]
Headings are incorrect

image

**Review by dkamburov — DISMISSED** [review]

File: en/components/grids_templates/cell-editing.md [hunk 4]
```diff
@@ -181,6 +181,85 @@ This code is used in the sample below which implements an [`IgxSelectComponent`]
...
+* `Enter`/ `Shift+Enter` navigation
+
+```typescript
+
+// should we hit enter, we use the grid's navigation tools to get to the bottom cell instead
+if (key == 13) {
+let thisRow = activeElem.row;
+const column = activeElem.column;
+const rowInfo = grid.dataView;
+this.grid.navigateTo(nextRow, column, (obj) => {
```
Comment (mddragnev) [inline en/components/grids_templates/cell-editing.md:228]: This snipped for `enter` key is missing some closing brackets and a line of code before the `if (key == 13)`
This is the line `let nextRow = this.getNextEditableRowIndex(thisRow, rowInfo, event.shiftKey);`

File: en/components/grids_templates/cell-editing.md [hunk 5]
```diff
@@ -181,6 +181,85 @@ This code is used in the sample below which implements an [`IgxSelectComponent`]
...
+
+<code-view style="height:550px" 
+           data-demos-base-url="{environment:demosBaseUrl}" 
+           iframe-src="{environment:demosBaseUrl}/grid/grid-editing-excel-style" alt="Angular Grid Excel Style Editing Example">
+</code-view>
+
+
+Main benefits of the above approach include:
+
+- Constant edit mode: typing while a cell is selected will immediately enter it in edit mode with the value typed, replacing the existing one
```
Comment (mddragnev) [inline en/components/grids_templates/cell-editing.md:258]: I think that changing this line to `Constant edit mode: typing while a cell is selected will immediately enter edit mode with the value typed, replacing the existing one` is a little more accurate.

**Review by mddragnev — CHANGES_REQUESTED** [review]

**Comment by mddragnev** [comment]
Also, putting some padding to the grid container will look better instead of adding margin on top and bottom. If you run your samples browser and docfx and go to the topic, observe the first sample of cell editing to get a better idea of what i am talking about.

File: en/components/grids_templates/cell-editing.md [hunk 6]
```diff
@@ -181,6 +181,85 @@ This code is used in the sample below which implements an [`IgxSelectComponent`]
...
+      }
+    }
+```
+
+* `Enter`/ `Shift+Enter` navigation
+
+```typescript
+
+// should we hit enter, we use the grid's navigation tools to get to the bottom cell instead
+if (key == 13) {
```
[the remark this answers is not in the document]
Comment (tishko0) [reply en/components/grids_templates/cell-editing.md]: added, please check if it looks okay now

File: en/components/grids_templates/cell-editing.md (hunk 4 above)
Comment (tishko0) [reply en/components/grids_templates/cell-editing.md]: should be looking good now

File: en/components/grids_templates/cell-editing.md (hunk 5 above)
Comment (tishko0) [reply en/components/grids_templates/cell-editing.md]: I have applied the changes, please review it again

File: en/components/grids_templates/cell-editing.md [hunk 7]
```diff
@@ -181,6 +181,80 @@ This code is used in the sample below which implements an [`IgxSelectComponent`]
...
+    (key >= 97 && key <= 122)){
+    // Number or Alphabet upper case or Alphabet lower case
+      const columnName = grid.getColumnByVisibleIndex(activeElem.column).field;
+      const cell = grid.getCellByColumn(activeElem.row, columnName);
+      if (cell && !grid.crudService.cellInEditMode) {
+        grid.crudService.enterEditMode(cell);
+        cell.editValue = event.key;
+      }
+     }
+    }
```
Comment (mddragnev) [inline en/components/grids_templates/cell-editing.md:213]: the last closing bracket is redundant.

File: en/components/grids_templates/cell-editing.md [hunk 8]
```diff
@@ -181,6 +181,80 @@ This code is used in the sample below which implements an [`IgxSelectComponent`]
...
+
+```typescript
+  if (key == 13) {
+      let thisRow = activeElem.row;
+      const column = activeElem.column;
+      const rowInfo = grid.dataView;
+      this.grid.navigateTo(nextRow, column, (obj) => {
+      obj.target.activate();
+  });
+  let nextRow = this.getNextEditableRowIndex(thisRow, rowInfo, event.shiftKey);
```
Comment (mddragnev) [inline en/components/grids_templates/cell-editing.md:226]: this part of the snippet is still not correct. It have many differences with the sample itself. Here you can copy the whole part of the snippet which is correct and formatted.
```
public keydownHandler(event) {
        const key = event.keyCode;
        const grid = this.grid;
        const activeElem = grid.navigation.activeNode;

        if (
            (key >= 48 && key <= 57) ||
            (key >= 65 && key <= 90) ||
            (key >= 97 && key <= 122)
        ) {
            // Number or Alphabet upper case or Alphabet lower case

            const columnName = grid.getColumnByVisibleIndex(activeElem.column).field;
            const cell = grid.getCellByColumn(activeElem.row, columnName);
            if (cell && !grid.crudService.cellInEditMode) {
                grid.crudService.enterEditMode(cell);
                cell.editValue = event.key;
            }
        }

        if (key == 13) {
            let thisRow = activeElem.row;
            const column = activeElem.column;
            const rowInfo = grid.dataView;

            let nextRow = this.getNextEditableRowIndex(thisRow, rowInfo, event.shiftKey);

            this.grid.navigateTo(nextRow, column, (obj) => {
                obj.target.activate();
                this.grid.clearCellSelection();
                this.cdr.detectChanges();
            });
        }
    }
```
You need only to leave the comments to the code as they are.

**Review by mddragnev — CHANGES_REQUESTED** [review]

**Review by mddragnev — APPROVED** [review]

**Pull request merged**